ggml-vulkan.cpp

```diff
@@ -734,17 +746,22 @@ static vk_buffer ggml_vk_create_buffer(ggml_backend_vk_context * ctx, size_t siz
 
     vk::PhysicalDeviceMemoryProperties mem_props = ctx->device.lock()->physical_device.getMemoryProperties();
 
-    uint32_t memory_type_index = UINT32_MAX;
+    uint32_t memory_type_index = -1;
 
-    for (uint32_t i = 0; i < mem_props.memoryTypeCount; ++i) {
-        vk::MemoryType memory_type = mem_props.memoryTypes[i];
-        if ((mem_req.memoryTypeBits & ((uint64_t)1 << i)) && (req_flags & memory_type.propertyFlags) == req_flags && mem_props.memoryHeaps[memory_type.heapIndex].size >= mem_req.size) {
-            memory_type_index = i;
-            break;
-        }
+    memory_type_index = find_properties(&mem_props, &mem_req, req_flags);
+    buf->memory_property_flags = req_flags;
+
+    // Failed to find memory type matching req_flags, but we can try again with fallback_flags if specified...
+    if (memory_type_index == -1 && fallback_flags && (
+        // ...as long as req_flags was either: 1) not DEVICE_LOCAL; or 2) DEVICE_LOCAL and device has UMA.
+        !(req_flags & vk::MemoryPropertyFlagBits::eDeviceLocal) ||
+         (req_flags & vk::MemoryPropertyFlagBits::eDeviceLocal && ctx->device.lock()->uma))
```

Comment:
I would prefer if the UMA logic stays in `ggml_vk_create_buffer_device` and this function just falls back if a fallback is given, without further checks.